### ISSUE TITLE
Pictures no longer appear

### ISSUE THREAD

**Issue by gitfreud** [HakanL/WkHtmlToPdf-DotNet#24]
Hello,
Since 1.3.0 version pictures (like .png) no longer appear. 
I checked after just changing runtime for 0.12.5 and it works fine.

**Comment by HakanL** [comment]
I see, that's strange. Can you push a PR with a unit test that shows the issue?

**Comment by gitfreud** [comment]
Yes, it's strange but the issue is already identified. I found [this:](https://github.com/wkhtmltopdf/wkhtmltopdf/issues/4832)

I confirm the workaround: if I remove width and height attributes, it works.

**Comment by HakanL** [comment]
Ok, thanks for confirming that it's an upstream/parent issue.

**Comment by gitfreud** [comment]
I made a mistake about the workaround. The bug exists with or without html attributes.

I opened a PR with an unit test. The test fails with wkhtmltopdf library 0.12.16 but it's good with 0.12.5 (win-64).

**Comment by HakanL** [comment]
Ah I see. I've added your unit test, thanks for your contribution. Unfortunately I think we're completely blocked by the issue in the parent project though.

**Comment by ramakrishnaindpro** [comment]
I have same problem using 0.12.6 (and no issue on 0.12.4), any updates on this issue?

**Comment by HakanL** [comment]
@ramakrishnaindpro It looks to be an issue with the parent wkhtmltopdf project, so it's not something that can be fixed in this .NET wrapper. See this issue: https://github.com/wkhtmltopdf/wkhtmltopdf/issues/4832

**Comment by Gunock** [comment]
Is there any chance this issue is with `--enable-local-file-access` parameter not being passes somwhere, or local file access is not set to be allowed when using this lib?
I found on stack overflow that from version 0.12.6 wkhtmltopdf does not allow for local files by default.

Here's link to this stack overflow lead: 
https://stackoverflow.com/questions/62315246/wkhtmltopdf-0-12-6-warning-blocked-access-to-file

Also changelog for version 0.12.6 supports this claim:
https://github.com/wkhtmltopdf/wkhtmltopdf/releases/0.12.6/

**Comment by HakanL** [comment]
@Gunock I don't think so, the error is when loading the library, before actually calling it.

**Comment by DanielVirta** [comment]
For me it works when first converting the image to a base64string and then including the image in the HTML in this form: src="data:image/png;base64,<data>".

**Comment by HakanL** [comment]
> Is there any chance this issue is with `--enable-local-file-access` parameter not being passed somewhere, or local file access is not set to be allowed when using this lib?
> I found on stack overflow that from version 0.12.6 wkhtmltopdf does not allow for local files by default.
> 
> Here's link to this stack overflow lead:
> https://stackoverflow.com/questions/62315246/wkhtmltopdf-0-12-6-warning-blocked-access-to-file
> 
> Also changelog for version 0.12.6 supports this claim:
> https://github.com/wkhtmltopdf/wkhtmltopdf/releases/0.12.6/

Maybe, but it seems to work if width/height is removed, so doesn't seem to be related to blocked access to local files (the example in the parent ticket was using an image hosted on a web server as well, so not local file).

**Comment by johguentner** [comment]
It looks like @Gunock findings are accurate.
Yesterday I had the same problem, while trying to access local files for css/fonts/images.

If you look at the doc changes for this parameter in wkhtmltopdf:
`--disable-local-file-access` is `(default)` in 0.12.6 and 
`--enable-local-file-acccess` is `(default)` in 0.12.5 
https://github.com/wkhtmltopdf/wkhtmltopdf/blob/0.12.6/docs/usage/wkhtmltopdf.txt
https://github.com/wkhtmltopdf/wkhtmltopdf/blob/0.12.5/docs/usage/wkhtmltopdf.txt

**Temporary Solution:**
Currently this can be solved by replacing the *.dll-files within the runtime-folder.
Just download the 0.12.5-Version for your system (https://github.com/wkhtmltopdf/wkhtmltopdf/releases/tag/0.12.5), install it, and copy the *.dll-File of the version to your runtime-folder
_(this is not technical advice, but only the way I solved it. Downgrading to an old version might lead to bugs in previous versions, or can have security issues, which are solved in the new version)_

**Comment by johguentner** [comment]
> It looks like @Gunock findings are accurate.
> Yesterday I had the same problem, while trying to access local files for css/fonts/images.
> 
> If you look at the doc changes for this parameter in wkhtmltopdf:
> `--disable-local-file-access` is `(default)` in 0.12.6 and
> `--enable-local-file-acccess` is `(default)` in 0.12.5
> https://github.com/wkhtmltopdf/wkhtmltopdf/blob/0.12.6/docs/usage/wkhtmltopdf.txt
> https://github.com/wkhtmltopdf/wkhtmltopdf/blob/0.12.5/docs/usage/wkhtmltopdf.txt
> 
> **Temporary Solution:**
> Currently this can be solved by replacing the *.dll-files within the runtime-folder.
> Just download the 0.12.5-Version for your system (https://github.com/wkhtmltopdf/wkhtmltopdf/releases/tag/0.12.5), install it, and copy the *.dll-File of the version to your runtime-folder
> _(this is not technical advice, but only the way I solved it. Downgrading to an old version might lead to bugs in previous versions, or can have security issues, which are solved in the new version)_

Since the initial issue declared by @gitfreud is not directly correlated to the local-file-access-issue and the disabled local-file-access is affecting not only images, I created a new issue, with the short description for a temporary soluation. Hope that's okay. (#31)

**Comment by mortb** [comment]
@johguentner Your suggestion is totally awesome! 
I upgraded to this nuget to get the 64-bit version and struggled with images that would not display, but adding this small snippet to the code solved my problems
` new ObjectSettings() { LoadSettings = new LoadSettings { BlockLocalFileAccess = false }, ... }`

Thanks for your effort

**Comment by johguentner** [comment]
For sure @mortb! You're welcome!

**Comment by Paul-SB** [comment]
Thanks for the tip BlockLocalFileAccess = false. This also fixed my problem that only the first conversion of an Outlook mail worked. The second conversion of another mail was always incorrect.

**Comment by HakanL** [comment]
It looks like this was related to the defaults changing in the parent library from version 0.12.5 to 0.12.6. BlockLocalFileAccess is the key here.

**Comment by xmontero** [comment]
Silly comment just in case someone is in my situation:

I've been struggling for hours without the images appearing in the PDF while they appeared if I opened the HTML in the browser.

It happened that I was opening an old "template" that pointed the images to `http://whatever`. While the browser did the call and got a 301 redirect to the very same image by TLS over `https://` the `wkhtmltopdf` just did not follow those redirects and the images appeared blank in the PDF.

v0.12.4 was not verbose about this.

I'm not reopening this bug, even not suggesting there's a bug. Only logging this to help people googling, as a google search about [wkhtmltopdf not rendering images] will probably lead to this "Pictures no longer appear" issue.

**Comment by Sicos1977** [comment]
https://github.com/Sicos1977/ChromeHtmlToPdf

**Comment by Jymmy3024** [comment]
> Xpost aquí: [wkhtmltopdf/wkhtmltopdf#4832 (comentario)](https://github.com/wkhtmltopdf/wkhtmltopdf/issues/4832#issuecomment-1193016846)
> 
> Comentario tonto por si alguien está en mi situación:
> 
> He estado luchando durante horas sin que aparecieran las imágenes en el PDF mientras aparecían si abría el HTML en el navegador.
> 
> Sucedió que estaba abriendo una "plantilla" antigua que apuntaba las imágenes `http://whatever`. Si bien el navegador hizo la llamada y obtuvo una redirección 301 a la misma imagen mediante TLS, `https://`simplemente `wkhtmltopdf`no siguió esas redirecciones y las imágenes aparecieron en blanco en el PDF. Cambiar las URL de las imágenes en la plantilla a las finales _evitando redirecciones_ resolvió el problema.
> 
> La v0.12.4 no fue detallada sobre esto.
> 
> No voy a reabrir este error, ni siquiera sugiero que haya un error. Registrar esto solo para ayudar a las personas a buscar en Google, ya que una búsqueda en Google sobre [wkhtmltopdf no representa imágenes] probablemente conducirá al problema "Las imágenes ya no aparecen".



> Xpost here: [wkhtmltopdf/wkhtmltopdf#4832 (comment)](https://github.com/wkhtmltopdf/wkhtmltopdf/issues/4832#issuecomment-1193016846)
> 
> Silly comment just in case someone is in my situation:
> 
> I've been struggling for hours without the images appearing in the PDF while they appeared if I opened the HTML in the browser.
> 
> It happened that I was opening an old "template" that pointed the images to `http://whatever`. While the browser did the call and got a 301 redirect to the very same image by TLS over `https://` the `wkhtmltopdf` just did not follow those redirects and the images appeared blank in the PDF. Changing the image URLs in the template to the final ones _avoiding any redirects_ solved the issue.
> 
> v0.12.4 was not verbose about this.
> 
> I'm not reopening this bug, even not suggesting there's a bug. Only logging this to help people googling, as a google search about [wkhtmltopdf not rendering images] will probably lead to this "Pictures no longer appear" issue.

How can I do it, I am using wkhtmltopdf in a proyect that is development in C#, aspx and javascript.

**Comment by Jymmy3024** [comment]
> @Jymmy3024 Perhaps you can proxy/cache the image in your web site so it's not a call into S3?

I don't know if is because the proyect is old, but it don't allow me to use localStorage

**Comment by HakanL** [comment]
> > @Jymmy3024 Perhaps you can proxy/cache the image in your web site so it's not a call into S3?
> 
> I don't know if is because the proyect is old, but it don't allow me to use localStorage

I don't think I can help you with that, just giving you suggestions on how to potentially solve the issue with your images. Bottom line is that it's best to have everything local when generating the PDFs, especially since it's also single-threaded.
